File: code/packages/typescript/lzss/tests/lzss.test.ts
import { describe, it, expect } from "vitest";
import {
  literal,
  match,
  encode,
  decode,
  compress,
  decompress,
  serialiseTokens,
  deserialiseTokens,
} from "../src/lzss.js";

// ─── Helpers ─────────────────────────────────────────────────────────────────

function enc(s: string): Uint8Array {
  return new TextEncoder().encode(s);
}

function rt(data: Uint8Array): Uint8Array {
  return decompress(compress(data));
}

function eq(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) return false;
  return true;
}

// ─── Spec vectors ─────────────────────────────────────────────────────────────

describe("encode — spec vectors", () => {
  it("empty input", () => {
    expect(encode(new Uint8Array())).toEqual([]);
  });

  it("single byte", () => {
    expect(encode(enc("A"))).toEqual([literal(65)]);
  });

  it("no repetition", () => {
    const tokens = encode(enc("ABCDE"));
    expect(tokens.length).toBe(5);
    expect(tokens.every((t) => t.kind === "literal")).toBe(true);
  });

  it("AABCBBABC", () => {
    const tokens = encode(enc("AABCBBABC"));
    expect(tokens.length).toBe(7);
    expect(tokens[6]).toEqual(match(5, 3));
  });

  it("ABABAB", () => {
    expect(encode(enc("ABABAB"))).toEqual([literal(65), literal(66), match(2, 4)]);
  });

  it("AAAAAAA (self-referential match)", () => {
    expect(encode(enc("AAAAAAA"))).toEqual([literal(65), match(1, 6)]);
  });
});

// ─── Encode properties ────────────────────────────────────────────────────────

describe("encode — properties", () => {
  it("match offset >= 1", () => {
    for (const tok of encode(enc("ABABABAB"))) {
      if (tok.kind === "match") expect(tok.offset).toBeGreaterThanOrEqual(1);
    }
  });

  it("match length >= min_match", () => {
    for (const tok of encode(enc("ABABABABABAB"))) {
      if (tok.kind === "match") expect(tok.length).toBeGreaterThanOrEqual(3);
    }
  });

  it("match offset within window_size", () => {
    const data = enc("ABCABCABCABC");
    for (const tok of encode(data, 4)) {
      if (tok.kind === "match") expect(tok.offset).toBeLessThanOrEqual(4);
    }
  });

  it("match length within max_match", () => {
    const data = new Uint8Array(100).fill(65);
    for (const tok of encode(data, 4096, 5)) {
      if (tok.kind === "match") expect(tok.length).toBeLessThanOrEqual(5);
    }
  });

  it("min_match large forces all literals", () => {
    const tokens = encode(enc("ABABAB"), 4096, 255, 100);
    expect(tokens.every((t) => t.kind === "literal")).toBe(true);
  });
});

// ─── Decode ───────────────────────────────────────────────────────────────────

describe("decode", () => {
  it("empty", () => {
    expect(decode([], 0)).toEqual(new Uint8Array());
  });

  it("single literal", () => {
    expect(decode([literal(65)], 1)).toEqual(new Uint8Array([65]));
  });

  it("overlapping match — AAAAAAA", () => {
    const result = decode([literal(65), match(1, 6)], 7);
    expect(result).toEqual(enc("AAAAAAA"));
  });

  it("ABABAB", () => {
    const result = decode([literal(65), literal(66), match(2, 4)], 6);
    expect(result).toEqual(enc("ABABAB"));
  });

  it("truncates to original_length", () => {
    const result = decode([literal(65), literal(66), literal(67)], 2);
    expect(result).toEqual(new Uint8Array([65, 66]));
  });

  it("returns all when -1", () => {
    const result = decode([literal(65), literal(66)]);
    expect(result).toEqual(new Uint8Array([65, 66]));
  });
});

// ─── Round-trip ───────────────────────────────────────────────────────────────

describe("round-trip", () => {
  const cases: [string, Uint8Array][] = [
    ["empty", new Uint8Array()],
    ["single", enc("A")],
    ["no repetition", enc("ABCDE")],
    ["all identical", enc("AAAAAAA")],
    ["ABABAB", enc("ABABAB")],
    ["AABCBBABC", enc("AABCBBABC")],
    ["hello world", enc("hello world")],
    ["ABC×100", enc("ABC".repeat(100))],
    ["full byte range", Uint8Array.from({ length: 256 }, (_, i) => i)],
    ["binary nulls", new Uint8Array([0, 0, 0, 255, 255])],
    ["repeated pattern", new Uint8Array(Array.from({ length: 300 }, (_, i) => i % 3))],
    ["long ABCDEF", enc("ABCDEF".repeat(500))],
    ["all zeros", new Uint8Array(1000)],
  ];

  for (const [name, data] of cases) {
    it(name, () => {
      expect(eq(rt(data), data)).toBe(true);
    });
  }
});

// ─── Wire format ─────────────────────────────────────────────────────────────

describe("wire format", () => {
  it("stores original_length in header", () => {
    const data = enc("hello");
    const compressed = compress(data);
    const view = new DataView(compressed.buffer);
    expect(view.getUint32(0, false)).toBe(5);
  });

  it("compress is deterministic", () => {
    const data = enc("hello world test");
    expect(compress(data)).toEqual(compress(data));
  });

  it("empty compresses to 8-byte header", () => {
    const c = compress(new Uint8Array());
    expect(c.length).toBe(8);
  });

  it("crafted large block_count is safe", () => {
    const bad = new Uint8Array(16);
    const view = new DataView(bad.buffer);
    view.setUint32(4, 0x40000000, false); // 2^30 blocks
    const result = decompress(bad);
    expect(result).toBeInstanceOf(Uint8Array);
  });
});

// ─── Compression effectiveness ───────────────────────────────────────────────

describe("compression effectiveness", () => {
  it("repetitive data compresses", () => {
    const data = enc("ABC".repeat(1000));
    expect(compress(data).length).toBeLessThan(data.length);
  });

  it("all-same byte compresses", () => {
    const data = new Uint8Array(10000).fill(0x42);
    const compressed = compress(data);
    expect(compressed.length).toBeLessThan(data.length);
    expect(eq(decompress(compressed), data)).toBe(true);
  });

  it("LZSS much smaller than raw on repetitive data", () => {
    const data = enc("ABCDEF".repeat(500)); // 3000 bytes
    expect(compress(data).length).toBeLessThan(data.length / 2);
  });
});

// ─── Serialise/deserialise symmetry ──────────────────────────────────────────

describe("serialise/deserialise", () => {
  it("symmetry for mixed tokens", () => {
    const tokens = [literal(65), literal(66), match(2, 4)];
    const bytes = serialiseTokens(tokens, 6);
    const [recovered, origLen] = deserialiseTokens(bytes);
    expect(origLen).toBe(6);
    expect(recovered).toEqual(tokens);
  });
});

File: code/packages/c/lzss/tests/lzss_test.c
/* Tests for the C lzss, using the iso_test.h harness. Vectors and round trips
 * are taken from the Rust crate's own tests. */
#include "iso_test.h"

#include <stdlib.h> /* malloc, free */
#include <string.h> /* memcmp, memset, strlen */

#include "lzss.h"

/* Encode with default params. */
static LzssToken *enc(const char *text, size_t *count) {
    LzssToken *toks = NULL;
    lzss_encode((const unsigned char *)text, strlen(text),
                LZSS_DEFAULT_WINDOW_SIZE, LZSS_DEFAULT_MAX_MATCH,
                LZSS_DEFAULT_MIN_MATCH, &toks, count);
    return toks;
}

static void check_round_trip(const unsigned char *data, size_t len) {
    unsigned char *comp = NULL, *back = NULL;
    size_t comp_len = 0, back_len = 0;
    ISO_CHECK_EQ_INT((int)lzss_compress(data, len, &comp, &comp_len),
                     (int)LZSS_OK);
    ISO_CHECK_EQ_INT((int)lzss_decompress(comp, comp_len, &back, &back_len),
                     (int)LZSS_OK);
    ISO_CHECK_EQ_UINT(back_len, len);
    if (back_len == len) {
        ISO_CHECK(len == 0 || memcmp(back, data, len) == 0);
    }
    free(comp);
    free(back);
}

int main(void) {
    /* Empty encode. */
    {
        size_t count = 99;
        LzssToken *toks = enc("", &count);
        ISO_CHECK_EQ_UINT(count, 0u);
        free(toks);
    }

    /* Single byte -> one literal. */
    {
        size_t count = 0;
        LzssToken *toks = enc("A", &count);
        ISO_CHECK_EQ_UINT(count, 1u);
        ISO_CHECK(!toks[0].is_match && toks[0].literal == 'A');
        free(toks);
    }

    /* No repetition -> all literals. */
    {
        size_t count = 0, i;
        LzssToken *toks = enc("ABCDE", &count);
        ISO_CHECK_EQ_UINT(count, 5u);
        for (i = 0; i < count; i++) {
            ISO_CHECK(!toks[i].is_match);
        }
        free(toks);
    }

    /* "AABCBBABC": 7 tokens, last is Match{offset:5, length:3}. */
    {
        size_t count = 0;
        LzssToken *toks = enc("AABCBBABC", &count);
        ISO_CHECK_EQ_UINT(count, 7u);
        if (count == 7) {
            ISO_CHECK(toks[6].is_match && toks[6].offset == 5 &&
                      toks[6].length == 3);
        }
        free(toks);
    }

    /* "ABABAB" -> [Lit A, Lit B, Match{2,4}]. */
    {
        size_t count = 0;
        LzssToken *toks = enc("ABABAB", &count);
        ISO_CHECK_EQ_UINT(count, 3u);
        if (count == 3) {
            ISO_CHECK(!toks[0].is_match && toks[0].literal == 'A');
            ISO_CHECK(!toks[1].is_match && toks[1].literal == 'B');
            ISO_CHECK(toks[2].is_match && toks[2].offset == 2 &&
                      toks[2].length == 4);
        }
        free(toks);
    }

    /* "AAAAAAA" -> [Lit A, Match{1,6}]. */
    {
        size_t count = 0;
        LzssToken *toks = enc("AAAAAAA", &count);
        ISO_CHECK_EQ_UINT(count, 2u);
        if (count == 2) {
            ISO_CHECK(!toks[0].is_match && toks[0].literal == 'A');
            ISO_CHECK(toks[1].is_match && toks[1].offset == 1 &&
                      toks[1].length == 6);
        }
        free(toks);
    }

    /* min_match large forces all literals. */
    {
        LzssToken *toks = NULL;
        size_t count = 0, i;
        lzss_encode((const unsigned char *)"ABABAB", 6, LZSS_DEFAULT_WINDOW_SIZE,
                    LZSS_DEFAULT_MAX_MATCH, 100, &toks, &count);
        for (i = 0; i < count; i++) {
            ISO_CHECK(!toks[i].is_match);
        }
        free(toks);
    }

    /* Matches stay within a small window. */
    {
        LzssToken *toks = NULL;
        size_t count = 0, i;
        lzss_encode((const unsigned char *)"ABCABCABCABC", 12, 8,
                    LZSS_DEFAULT_MAX_MATCH, LZSS_DEFAULT_MIN_MATCH, &toks,
                    &count);
        for (i = 0; i < count; i++) {
            if (toks[i].is_match) {
                ISO_CHECK(toks[i].offset <= 8);
            }
        }
        free(toks);
    }

    /* Match length bounded by max_match. */
    {
        unsigned char aaa[100];
        LzssToken *toks = NULL;
        size_t count = 0, i;
        memset(aaa, 'A', sizeof aaa);
        lzss_encode(aaa, sizeof aaa, LZSS_DEFAULT_WINDOW_SIZE, 5,
                    LZSS_DEFAULT_MIN_MATCH, &toks, &count);
        for (i = 0; i < count; i++) {
            if (toks[i].is_match) {
                ISO_CHECK(toks[i].length <= 5);
            }
        }
        free(toks);
    }

    /* decode vectors. */
    {
        unsigned char *out = NULL;
        size_t out_len = 0;
        LzssToken empty_lit = {0, 'A', 0, 0};
        LzssToken tokens[2];
        /* empty */
        lzss_decode(NULL, 0, 1, 0, &out, &out_len);
        ISO_CHECK_EQ_UINT(out_len, 0u);
        free(out);
        /* single literal */
        lzss_decode(&empty_lit, 1, 1, 1, &out, &out_len);
        ISO_CHECK(out_len == 1 && out[0] == 'A');
        free(out);
        /* overlapping match: Lit A + Match{1,6} -> AAAAAAA */
        tokens[0].is_match = 0;
        tokens[0].literal = 'A';
        tokens[0].offset = 0;
        tokens[0].length = 0;
        tokens[1].is_match = 1;
        tokens[1].literal = 0;
        tokens[1].offset = 1;
        tokens[1].length = 6;
        lzss_decode(tokens, 2, 1, 7, &out, &out_len);
        ISO_CHECK_EQ_UINT(out_len, 7u);
        ISO_CHECK(out_len == 7 && memcmp(out, "AAAAAAA", 7) == 0);
        free(out);
    }

    /* Round trips (text + binary). */
    {
        const char *texts[] = {"",         "A",        "ABCDE",
                               "AAAAAAA",  "ABABABAB", "AABCBBABC",
                               "hello world hello world", "the quick brown fox"};
        size_t i;
        for (i = 0; i < 8; i++) {
            check_round_trip((const unsigned char *)texts[i], strlen(texts[i]));
        }
    }
    {
        unsigned char all256[256];
        unsigned char reps[3000];
        size_t i;
        for (i = 0; i < 256; i++) {
            all256[i] = (unsigned char)i;
        }
        check_round_trip(all256, 256);
        for (i = 0; i < 3000; i++) {
            reps[i] = (unsigned char)"ABC"[i % 3];
        }
        check_round_trip(reps, 3000);
    }

    /* Repetitive data compresses below its original size. */
    {
        unsigned char *data = malloc(10000);
        unsigned char *comp = NULL;
        size_t comp_len = 0;
        memset(data, 'A', 10000);
        lzss_compress(data, 10000, &comp, &comp_len);
        ISO_CHECK(comp_len < 10000);
        free(data);
        free(comp);
    }

    /* Malformed decompress must not crash (bad offsets, capped blocks). */
    {
        unsigned char bad[] = {0,   0,   0,   8,   /* orig_len 8 */
                               0,   0,   0,   99,  /* block_count 99 (capped) */
                               0x01, 0xFF, 0xFF, 5, /* flag: token0 match; off 65535 (>output) */
                               0x00, 66};           /* flag: token0 literal 'B' */
        unsigned char *out = NULL;
        size_t out_len = 0;
        ISO_CHECK_EQ_INT((int)lzss_decompress(bad, sizeof bad, &out, &out_len),
                         (int)LZSS_OK);
        free(out);
    }

    return ISO_TEST_RESULT();
}
